sycl/doc/CompileTimeProperties.md

```diff
@@ -0,0 +1,813 @@
+`properties` member variable to a named kernel function object:
+
+```
+using sycl;
+using sycl::ext::oneapi;
+
+class MyKernel {
+ public:
+  void operator()() {/* ... */}
+
+  static constexpr auto properties =
+    property_list{sub_group_size_v<32>, device_has_v<aspect::fp16>};
+};
+
+void foo(handler &cgh) {
+  MyKernel k;
+  cgh.single_task(k);
+}
+```
+
+Internally, the headers lower both cases to a wrapper class that expresses the
+properties as an internal C++ attribute, and the `operator()` of this class
+becomes the "top level" kernel function that is recognized by the front-end.
+
+```
+template<typename KernelType, typename PropertyListT>
+class KernelSingleTaskWrapper;
+
+// Partial specialization to make PropertyListT visible as a parameter pack
+// of properties.
+template<typename KernelType, typename ...Props>
+class KernelSingleTaskWrapper<KernelType, property_list<Props...>> {
+  KernelType k;
+
+ public:
+  KernelSingleTaskWrapper(KernelType k) : k(k) {}
+
+#ifdef __SYCL_DEVICE_ONLY__
+  __attribute__((sycl_kernel))
+  [[__sycl_detail__::add_ir_function_attributes(
+    Props::meta_name..., Props::meta_value...
+    )]]
+#endif
+  void operator()() {k();}
+};
+```
+
+Although the DPC++ headers only use the
+`[[__sycl_detail__::add_ir_function_attributes()]]` attribute on the definition
+of a kernel function as shown above, the front-end recognizes it for any
+function definition.  The front-end adds one LLVM IR function attribute for
+each property in the list.  For example, this can be done by calling
+[`Function::addFnAttr(StringRef, StringRef)`][9].  As before, the IR attributes
+are added as strings, so the front-end must convert the property value to a
+string if it is not already a string.
+
+[9]: <https://llvm.org/doxygen/classllvm_1_1Function.html#ae7b919df259dce5480774e656791c079>
+
+**NOTE**: The intention is to replace the existing member functions like
+`handler::kernel_single_task()` with wrapper classes like
+`KernelSingleTaskWrapper`.  We believe this will not cause problems for the
+device compiler front-end because it recognizes kernel functions via the
+`__attribute__((sycl_kernel))` attribute, not by the name
+`handler::kernel_single_task()`.
+
+
+## Properties on a non-global variable type
+
+Another use of compile-time properties is with types that are used to define
+non-global variables.  An example of this is the proposed `annotated_ptr`
+class.
+
+```
+namespace sycl::ext::oneapi {
+
+template <typename T, typename PropertyListT = property_list_t<>>
+class annotated_ptr {
+  T *ptr;
+ public:
+  annotated_ptr(T *p) : ptr(p) {}
+};
+
+} // namespace sycl::ext::oneapi
+```
+
+where an example use looks like:
+
+```
+using sycl::ext::oneapi;
+
+void foo(int *p) {
+  annotated_ptr<int
+    property_list_t<
+      foo::value_t,
+      bar::value_t<32>>>
+    aptr(p);
+}
+```
+
+We again implement the property list in the header via a C++ attribute, where
```

Comment:
Here as well.

Reply:
Done in 403e6c964c4539197fc186aee9a9ce2ff271ebe7